### PR TITLE
Make menu item edit and tree view in admin

### PR DESCRIPTION
**Pull request by CheAlex**
- make menu item tree view in admin
- make menu item admin edit

### CODE REVIEW

File: Admin/AdminExtension/MenuAdminExtension.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+use Grossum\MenuBundle\Entity\EntityManager\BaseMenuItemManager;
+
+use Doctrine\ORM\EntityManager;
+
+class MenuAdminExtension extends AdminExtension
+{
+    /**
+     * @var EntityManager
+     */
+    private $entityManager;
```
Comment (malavsky) [inline Admin/AdminExtension/MenuAdminExtension.php]: make it protected

File: Admin/AdminExtension/MenuAdminExtension.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    protected $menuItemManager;
+
+    /**
+     * @param BaseMenuItemManager $menuItemManager
+     * @param EntityManager $entityManager
+     */
+    public function __construct(BaseMenuItemManager $menuItemManager, EntityManager $entityManager)
+    {
+        $this->menuItemManager = $menuItemManager;
+        $this->entityManager = $entityManager;
```
Comment (malavsky) [inline Admin/AdminExtension/MenuAdminExtension.php]: In MenuItemManager, you already have entity manager. Use em from item manager.

File: Admin/AdminExtension/MenuAdminExtension.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function prePersist(AdminInterface $admin, $object)
+    {
+        /* @var $object BaseMenu */
+
```
Comment (malavsky) [inline Admin/AdminExtension/MenuAdminExtension.php]: remove new line.

File: Admin/MenuItemAdmin.php [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+{
+    /**
+     * {@inheritdoc}
+     */
+    protected $parentAssociationMapping = 'menu';
+
+    /**
+     * @var BaseMenuItemManager
+     */
+    private $menuItemManager;
```
Comment (malavsky) [inline Admin/MenuItemAdmin.php]: make it protected

File: Admin/MenuItemAdmin.php [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+    /**
+     * @var MenuManager
+     */
+    private $menuMaster;
```
Comment (malavsky) [inline Admin/MenuItemAdmin.php]: make it protected

File: Admin/AdminExtension/MenuAdminExtension.php (hunk 1 above)
Comment (CheAlex) [inline Admin/AdminExtension/MenuAdminExtension.php]: Why it should be protected?

**Comment by aivus** [comment]
Also, all `TODO`s should be fixed and test code removed before merge

File: Entity/BaseMenuItem.php [hunk 6]
```diff
@@ -126,8 +158,7 @@ public function setEntityIdentifier($entityIdentifier)
     }
 
     /**
-     * Get entityIdentifier
-     *
+    *
```
Comment (aivus) [inline Entity/BaseMenuItem.php]: Missed whitespace

File: Admin/Form/EventListener/AddEntityIdentifierFieldSubscriber.php [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+     * @param FormEvent $event
+     */
+    public function onPreSetData(FormEvent $event)
+    {
+        $form = $event->getForm();
+
+        /* @var $menuItem BaseMenuItem */
+        $menuItem = $event->getData();
+
+        if ($menuItem xor $menuItem->getEntityClass()) {
```
Comment (aivus) [inline Admin/Form/EventListener/AddEntityIdentifierFieldSubscriber.php]: `$menuItem ->getUrl()` instead of `$menuItem` I think

Comment (aivus) [inline Admin/Form/EventListener/AddEntityIdentifierFieldSubscriber.php]: @cowboy4ik89 could you check this expression again?

File: Admin/Form/EventListener/AddEntityIdentifierFieldSubscriber.php [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+    /**
+     * @param FormEvent $event
+     */
+    public function onPreSubmit(FormEvent $event)
+    {
+        $data = $event->getData();
+        $form = $event->getForm();
+
+        if (isset($data['entityClass']) xor $data['entityClass']) {
```
Comment (aivus) [inline Admin/Form/EventListener/AddEntityIdentifierFieldSubscriber.php]: There is something strange.
@cowboy4ik89 could you check this expression again?

File: Admin/Form/EventListener/AddEntityIdentifierFieldSubscriber.php [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    {
+        $choices = $this->menuMaster->getMenuHandler($entityClass)->getIdentifierList();
+
+        if ($form->has('entityIdentifier')) {
+            $form->remove('entityIdentifier');
+        }
+
+        $form->add(
+            'entityIdentifier',
+            'choice',
```
Comment (aivus) [inline Admin/Form/EventListener/AddEntityIdentifierFieldSubscriber.php]: I think, we can take type and options from exists form.
Like a:

```
$entityIdentifierField = $form->get('entityIdentifier');
$config = $entityIdentifierField->getConfig()->getOptions(); // merge this options with new choices
$form->add($entityIdentifierField->getName(), $entityIdentifierField->getType(), $config);
```

File: Admin/MenuAdmin.php [hunk 10]
```diff
@@ -25,4 +28,40 @@ protected function configureListFields(ListMapper $listMapper)
...
+                    ]
+                )
+            ]
+        );
+
+        $menu->addChild(
+            $this->trans('grossum_menu.admin.side_menu.link_items_list'),
+            [
+                'uri' => $this->generateUrl(
+                    'grossum_menu.admin.menu|grossum_menu.admin.menu_item.list',
```
Comment (aivus) [inline Admin/MenuAdmin.php:59]: Is it ok to have `|` here?

File: Admin/MenuItemAdmin.php [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+use Sonata\AdminBundle\Admin\Admin;
+use Sonata\AdminBundle\Datagrid\ListMapper;
+use Sonata\AdminBundle\Form\FormMapper;
+use Sonata\AdminBundle\Route\RouteCollection;
+
+use Grossum\MenuBundle\Entity\EntityManager\BaseMenuItemManager;
+use Grossum\MenuBundle\Manager\MenuManager;
+use Grossum\MenuBundle\Form\EventListener\AddEntityIdentifierFieldSubscriber;
+use Symfony\Component\Form\Extension\Core\Type\ChoiceType;
```
Comment (aivus) [inline Admin/MenuItemAdmin.php]: Move it on top of the list

File: Admin/MenuItemAdmin.php [hunk 12]
```diff
@@ -0,0 +1,137 @@
+<?php
+
+namespace Grossum\MenuBundle\Admin;
+
+use Grossum\MenuBundle\Form\Type\MenuEntityClassType;
```
Comment (aivus) [inline Admin/MenuItemAdmin.php]: Move it down

File: Controller/MenuItemAdminController.php [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+     * @throws \Twig_Error_Runtime
+     */
+    public function treeAction(Request $request)
+    {
+        $menuId = $request->get($this->admin->getParent()->getIdParameter());
+        $menu = $this
+            ->get('grossum_menu.menu.entity.manager')
+            ->getRepository()
+            ->find($menuId);
+        /* @var $menu BaseMenu */
```
Comment (aivus) [inline Controller/MenuItemAdminController.php]: Move it up

File: Controller/MenuItemAdminController.php [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+    public function saveTreeAction(Request $request)
+    {
+        $tree = $request->request->get('tree');
+
+        $menuId = $request->get($this->admin->getParent()->getIdParameter());
+        $menu = $this
+            ->get('grossum_menu.menu.entity.manager')
+            ->getRepository()
+            ->find($menuId);
+        /* @var $menu BaseMenu */
```
Comment (aivus) [inline Controller/MenuItemAdminController.php]: Move it up

File: Entity/Repository/BaseMenuItemRepository.php [hunk 15]
```diff
@@ -2,8 +2,95 @@
...
+            ->andWhere(
+                $qb->expr()->notIn('menu_item.id', ':except')
+            )
+            ->setParameter('menu', $menuId)
+            ->setParameter(
+                'except',
+                array_map(
+                    function ($menuItem) {
+                        /* @var $menuItem BaseMenuItem */
+                        return $menuItem->getId();
```
Comment (aivus) [inline Entity/Repository/BaseMenuItemRepository.php]: Is it required?
Could you test it without this `array_map` call?

File: Form/DataTransformer/IdToMenuItem.php [hunk 16]
```diff
@@ -0,0 +1,64 @@
+<?php
+
+namespace Grossum\MenuBundle\Form\DataTransformer;
+
+use Symfony\Component\Form\DataTransformerInterface;
+use Symfony\Bridge\Doctrine\Form\ChoiceList\IdReader;
+
+use Doctrine\Common\Persistence\ObjectManager;
+
+class IdToMenuItem implements DataTransformerInterface
```
Comment (aivus) [inline Form/DataTransformer/IdToMenuItem.php]: Please, rename it to `IdToMenuItemTransformer`

File: Form/EventListener/AddEntityIdentifierFieldSubscriber.php [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+
+    /**
+     * @param FormEvent $event
+     */
+    public function onPreSubmit(FormEvent $event)
+    {
+        $form = $event->getForm();
+        $data = $event->getData();
+
+        if (!$data || !isset($data['entityClass']) || !$data['entityClass']) {
```
Comment (aivus) [inline Form/EventListener/AddEntityIdentifierFieldSubscriber.php]: You can remove `!$data` here

File: Resources/config/doctrine/MenuItem.orm.yml.skeleton [hunk 18]
```diff
@@ -9,3 +9,28 @@ Application\Grossum\MenuBundle\Entity\MenuItem:
             id: true
             generator:
                 strategy: AUTO
+#    @todo: We wrote associations here because Gedmo-Extension can see associations only in config-file - it doesn't use
```
Comment (aivus) [inline Resources/config/doctrine/MenuItem.orm.yml.skeleton]: Please, remove this todo

File: Resources/config/services.yml [hunk 19]
```diff
@@ -1,10 +1,19 @@
+parameters:
+    grossum_menu_tree_depth: 32
+
 services:
     grossum_menu.menu.entity.manager:
-        class: %grossum_menu.menu.manager.class%
+        class: %grossum_menu.menu.entity.manager.class%
```
Comment (aivus) [inline Resources/config/services.yml]: I think, we should rename it as `grossum_menu.entity_manager.base_menu.class` or something like this

Comment (aivus) [inline Resources/config/services.yml]: It relates to service name too

File: Admin/MenuAdmin.php (hunk 10 above)
Comment (CheAlex) [inline Admin/MenuAdmin.php:59]: Yes:
https://github.com/sonata-project/SonataPageBundle/blob/master/Admin/PageAdmin.php#L299

File: Entity/Repository/BaseMenuItemRepository.php (hunk 15 above)
Comment (CheAlex) [inline Entity/Repository/BaseMenuItemRepository.php]: Yes, it work without `array_map`

**Comment by CheAlex** [comment]
This pull request depends on this https://github.com/GrossumUA/MenuBundle/pull/6 request

**Pull request merged**